arch/arm/src/stm32h7/stm32_fdcan_sock.c

```diff
@@ -0,0 +1,2422 @@
+      return;
+    }
+
+  /* Check for message lost; count an error */
+
+  if ((*rxfns & FDCAN_RXF0S_RF0L) != 0)
+    {
+      NETDEV_RXERRORS(&priv->dev);
+    }
+
+  /* Check number of elements available (fill level) */
+
+  const uint8_t n_elem = (*rxfns & FDCAN_RXF0S_F0FL);
+
+  if (n_elem == 0)
+    {
+      nerr("RX interrupt but 0 frames available");
+      return;
+    }
+
+  struct rx_fifo_s *rf = NULL;
+
+  while ((*rxfns & FDCAN_RXF0S_F0FL) > 0)
+    {
+      /* Copy the frame from message RAM */
+
+      const uint8_t index = (*rxfns & FDCAN_RXF0S_F0GI) >>
+                             FDCAN_RXF0S_F0GI_SHIFT;
+
+      rf = &priv->rx[index];
+
+      /* Read the frame contents */
+
+#ifdef CONFIG_NET_CAN_CANFD
+      if (rf->header.fdf) /* CAN FD frame */
+        {
+          struct canfd_frame *frame = (struct canfd_frame *)priv->rx_pool;
+
+          if (rf->header.id.xtd)
+            {
+              frame->can_id  = CAN_EXT_MASK & rf->header.id.extid;
+              frame->can_id |= CAN_EXT_FLAG;
+            }
+          else
+            {
+              frame->can_id = CAN_STD_MASK & rf->header.id.stdid;
+            }
+
+          if (rf->header.id.rtr)
+            {
+              frame->can_id |= CAN_RTR_FLAG;
+            }
+
+          frame->len = can_dlc_to_len[rf->header.dlc];
+
+          uint32_t *frame_data_word = (uint32_t *)&frame->data[0];
+
+          for (int i = 0; i < (frame->len + 4 - 1) / 4; i++)
+            {
+              frame_data_word[i] = rf->data[i].word;
+            }
+
+          /* Acknowledge receipt of this FIFO element */
+
+          putreg32(index, rxfna);
+
+          /* Copy the buffer pointer to priv->dev
+           * Set amount of data in priv->dev.d_len
+           */
+
+          priv->dev.d_len = sizeof(struct canfd_frame);
+          priv->dev.d_buf = (uint8_t *)frame;
+        }
+      else /* CAN 2.0 Frame */
+#endif
+        {
+          struct can_frame *frame = (struct can_frame *)priv->rx_pool;
+
+          if (rf->header.id.xtd)
+            {
+              frame->can_id  = CAN_EXT_MASK & rf->header.id.extid;
+              frame->can_id |= CAN_EXT_FLAG;
+            }
+          else
+            {
+              frame->can_id = CAN_STD_MASK & rf->header.id.stdid;
+            }
+
+          if (rf->header.id.rtr)
+            {
+              frame->can_id |= CAN_RTR_FLAG;
+            }
+
+          frame->can_dlc = rf->header.dlc;
+
+          *(uint32_t *)&frame->data[0] = rf->data[0].word;
+          *(uint32_t *)&frame->data[4] = rf->data[1].word;
+
+          /* Acknowledge receipt of this FIFO element */
+
+          putreg32(index, rxfna);
+
+          /* Copy the buffer pointer to priv->dev
+           * Set amount of data in priv->dev.d_len
+           */
+
+          priv->dev.d_len = sizeof(struct can_frame);
+          priv->dev.d_buf = (uint8_t *)frame;
+        }
+
+      /* Send to socket interface */
+
+      can_input(&priv->dev);
```

Comment:
Does it work with DEBUGASSERT enabled ?
It seems to me that this is the same case as in https://github.com/apache/incubator-nuttx/issues/5142
This interface should not be called from the interrupt level

Reply:
Interesting.  @PetervdPerk-NXP have you found a solution or workaround to this issue with your work on NXP's boards?

Reply:
I think that the only solution is to delegate the receiver logic to work queue. This is how stm32 SocketCAN is done (and all others network drivers)

Reply:
That sounds correct.  It looks like Peter found the real-time performance of that to be unacceptable, though.  I'll see if I can do some profiling / timing to verify this doesn't add too much latency to the system (note that we're aiming to use the CAN bus for command and control of actuators on a large VTOL platform, so low latency is critical to our specific application).

To be clear, the (possible) latency issue is more my problem than a NuttX problem.  A work queue seems like the right call for this PR, so I'll make that change before we merge this in.

Reply:
It was mostly related to dropped CAN frames since the delay of the workqueue was so high that all my RX mailboxes were full before the workqueue logic could even empty them.

Maybe this was regression in the workqueue logic, since this was tested with low system load and different priorities for the workqueue, @JacobCrabill might re-asses this.

As alternative I was thinking to do a kthread just like they did for the Tiva CAN driver
https://github.com/apache/incubator-nuttx/blob/7f80d4ded69dd7da752d3302c1dbd5c246392653/arch/arm/src/tiva/common/tiva_can.c#L410-L414

But I didn't have time yet to test it and assess the peformance.